zmq.cpp

```diff
@@ -68,18 +81,53 @@ void LoadBalancer::run(){
         }
         //handle clients request
         if (items[1].revents & ZMQ_POLLIN){
+            // The client request is a multi-part ZMQ message, we have to check every frame and be sure the multi-part message frame
+            // is composed as we wish, otherwise the multi-part message may be shifted unexpectedly.
+
+            int more;
+            size_t more_size = sizeof (more);
+
             //first we get the client id
             zmq::message_t msg_identity;
             clients.recv(&msg_identity);
-            {
-                //then an empty frame
-                std::string empty = z_recv(clients);
-                assert(empty.size() == 0);
+            // Are there more frames coming?
+            clients.getsockopt(ZMQ_RCVMORE, &more, &more_size);
+            assert(more == 1);
+
+            // We're waiting for an empty frame...
+            // If no more frames, we reply immediately
+            if (more == 0) {
+                bool should_purge = false;
+                purge_and_reply(clients, more, should_purge);
+                continue;
+            }
+
+            //then an empty frame, there should be another frame follows up...
+            std::string empty = z_recv(clients);
```

Comment:
```suggestion
            const auto empty_frame = z_recv(clients);
```
It is weird to have empty.empty() afterwards and const is always welcome.